R/join_keys.R

```diff
@@ -0,0 +1,599 @@
+
+  checkmate::assert_list(new_join_keys, types = c("JoinKeys"), min.len = 1)
+
+  result <- join_keys_obj
+
+  for (jk in new_join_keys) {
+    for (dataset_1 in names(jk)) {
+      for (dataset_2 in names(jk[[dataset_1]])) {
+        result[[dataset_1]][[dataset_2]] <- jk[[dataset_1]][[dataset_2]]
+      }
+    }
+  }
+  logger::log_trace("JoinKeys keys merged.")
+  return(result)
+}
+
+#' Prints `JoinKeys`.
+#'
+#' @inheritParams base::print
+#' @return the `x` parameter
+#'
+#' @export
+print.JoinKeys <- function(x, ...) {
+  check_ellipsis(...)
+  keys_list <- x
+  my_parents <- parents(keys_list)
+  class(keys_list) <- "list"
+  if (length(keys_list) > 0) {
+    cat(sprintf(
+      "A JoinKeys object containing foreign keys between %s datasets:\n",
+      length(keys_list)
+    ))
+    # Hide parents
+    attr(keys_list, "__parents__") <- NULL
+    print.default(keys_list)
+  } else {
+    cat("An empty JoinKeys object.")
+  }
+  invisible(x)
+}
+
+# -----------------------------------------------------------------------------
+#
+#
+# Helpers (non-exported)
+#
+
+#' Internal constructor
+#'
+#' @return an empty `JoinKeys` list
+#'
+#' @keywords internal
+new_join_keys <- function() {
+  result <- list()
+  class(result) <- c("JoinKeys", "list")
+  result
+}
+
+#' Get value of a single relationship pair
+#'
+#' @param join_keys_obj (`JoinKeys`) object that holds the relationship keys.
+#' @param dataset_1 (`character(1)`) one of the datasets to retrieve keys (
+#' order of the datasets is irrelevant).
+#' @param dataset_2 (`character(1)`) the other dataset to retrieve keys (the
+#' order of the datasets is irrelevant).
+#'
+#' @return Character vector with keys or (if one of the datasets is omitted) a
+#' list of relationship pairs. If both datasets are omitted it returens the
+#' `JoinKeys` object
+#'
+#' @keywords internal
+get_join_key <- function(join_keys_obj, dataset_1 = NULL, dataset_2 = NULL) {
+  if (checkmate::test_integerish(dataset_1, len = 2)) {
+    # if dataset_1 is an index integet vector, then return itself
+    #  trick to make the following work: join_keys(jk)["ds1", "ds2"] <- "key"
+    return(join_keys_obj)
+  }
+  checkmate::assert_string(dataset_1, null.ok = TRUE)
+  if (missing(dataset_2)) {
+    # protection if dataset_2 is passed through by a function
+    dataset_2 <- NULL
+  }
+  checkmate::assert_string(dataset_2, null.ok = TRUE)
+  assert_join_keys(join_keys_obj)
+
+  if (is.null(dataset_1) && is.null(dataset_2)) {
+    return(join_keys_obj)
+  }
+  if (is.null(dataset_2)) {
+    return(join_keys_obj[[dataset_1]])
+  }
+  if (is.null(dataset_1)) {
+    return(join_keys_obj[[dataset_2]])
+  }
+
+  result <- join_keys_obj[[dataset_1]][[dataset_2]]
+
+  if (is.null(result)) {
+    return(character(0))
+  }
+  result
+}
+
+#' Internal assignment of value to a JoinKeys object
+#'
+#' @inheritParams join_keys
+#'
+#' @keywords internal
+add_key <- function(join_keys_obj, dataset_1, dataset_2 = dataset_1, value) {
```

Comment:
I think it is not needed. One can use `modifyList`, `append.list`, `c.list`

Reply:
I agree, especially now!

Added a `[[<-.JoinKey` method that adds some protections on the symmetry of pairs. 

It is very difficult to add only one side of the relationship.